sql/hive/src/test/scala/org/apache/spark/sql/sources/BucketedReadSuite.scala

```diff
@@ -59,6 +61,159 @@ class BucketedReadSuite extends QueryTest with SQLTestUtils with TestHiveSinglet
     }
   }
 
+  // To verify if the bucket pruning works, this function checks two conditions:
+  //   1) Check if the pruned buckets (before filtering) are empty.
+  //   2) Verify the final result is the same as the expected one
+  private def checkPrunedAnswers(
+      bucketSpec: BucketSpec,
+      bucketValues: Seq[Integer],
+      bucketedDataFrame: DataFrame,
+      expectedAnswer: DataFrame): Unit = {
+
+    val BucketSpec(numBuckets, bucketColumnNames, _) = bucketSpec
+    // Limit: bucket pruning only works when the bucket column has one and only one column
+    assert(bucketColumnNames.length == 1)
+    val bucketColumnIndex = bucketedDataFrame.schema.fieldIndex(bucketColumnNames.head)
+    val bucketColumn = bucketedDataFrame.schema.toAttributes.head
+    val matchedBuckets = new BitSet(numBuckets)
+    bucketValues.foreach { value =>
+      matchedBuckets.set(DataSourceStrategy.getBucketId(bucketColumn, numBuckets, value))
+    }
+
+    // Filter could hide the bug in bucket pruning. Thus, skipping all the filters
+    val rdd = bucketedDataFrame.queryExecution.executedPlan.find(_.isInstanceOf[PhysicalRDD])
+    assert(rdd.isDefined)
+
+    val checkBucketId = rdd.get.execute().map(_.copy()).mapPartitions(iter => {
+      iter.map(row =>
+        DataSourceStrategy.getBucketId(
+          bucketColumn, numBuckets, row.get(bucketColumnIndex, bucketColumn.dataType)))})
+    // Check if all the returned rows are from the non-pruned buckets
+    assert(checkBucketId.collect().forall(matchedBuckets.get))
+
+    checkAnswer(
+      expectedAnswer
+        .orderBy(expectedAnswer.logicalPlan.output.map(attr => Column(attr)) : _*),
+      bucketedDataFrame
+        .orderBy(bucketedDataFrame.logicalPlan.output.map(attr => Column(attr)) : _*))
+  }
+
+  test("read partitioning bucketed tables with bucket pruning filters") {
+    withTable("bucketed_table") {
+      val numBuckets = 50
+      val bucketSpec = BucketSpec(numBuckets, Seq("j"), Nil)
+      // json does not support predicate push-down, and thus json is used here
+      df.write
+        .format("json")
+        .partitionBy("i")
+        .bucketBy(numBuckets, "j")
+        .saveAsTable("bucketed_table")
+
+      for (j <- 0 until 13) {
+        // Case 1: EqualTo
+        checkPrunedAnswers(
+          bucketSpec,
+          bucketValues = j :: Nil,
+          hiveContext.table("bucketed_table")
+            .select("i", "j", "k").filter($"j" === j),
+          df.select("i", "j", "k").filter($"j" === j))
```

Comment:
the order of fields in `df` is always `i, j, k` right? I think we only need to re-order the read-back hive table